### ISSUE TITLE
Support for the official AWS SDK for Rust?

### ISSUE THREAD

**Issue by dcormier** [zenlist/serde_dynamo#9]
Any interest in supporting [the official AWS SDK for Rust](https://aws.amazon.com/blogs/developer/a-new-aws-sdk-for-rust-alpha-launch/)? Specifically, [its implementation of `AttributeValue`](https://github.com/awslabs/aws-sdk-rust/blob/ef1010197385eb39b3ed05da618f137cbde4f63a/sdk/dynamodb/src/model.rs#L6210-L6277)?

If so, thoughts on how to approach it? As a [feature](https://doc.rust-lang.org/cargo/reference/features.html), maybe? Perhaps there should be two mutually exclusive feature flags: `rusoto` and `aws`, with `rusoto` being the default. If `aws` is enabled, then it serializes/deserializes to/from the AWS `AttributeValue` type instead of `rusoto_dynamodb::AttributeValue`?

**Comment by bryanburgers** [comment]
You’re on top of it! Wow!

Yes, we discussed it on Friday and I was hoping to look into it a bit on Monday.

**Comment by bryanburgers** [comment]
My first thought was to use features, but not have a default.

`serde_dynamo::rusoto::from_item` or `serde_dynamo::aws_sdk::from_item`.

I’m not sure, though.

**Comment by dcormier** [comment]
> My first thought was to use features, but not have a default.

That's an option. Though, if you have `rusoto` be the default feature, users upgrading wouldn't have to do anything.

If I were an unknowing user upgrading and ran into a breaking change like that, I wouldn't mind as long as the solution were prominently documented. Maybe at the top of both the readme (in case they go to crates.io or the repo) and the package docs (in case they go there).

> `serde_dynamo::rusoto::from_item` or `serde_dynamo::aws_sdk::from_item`.

Could go that route, for sure. And have the features simply change which is called by `serde_dynamo::from_item`. Could even have `serde_dynamo::rusoto` and `serde_dynamo::aws_sdk` be just `pub(crate)` so the actual `pub` API stays pretty simple. Just a thought.

**Comment by bryanburgers** [comment]
I pushed a super-preview branch – [3.0.0-alpha.0](https://github.com/zenlist/serde_dynamo/tree/3.0.0-alpha.0) – that worked for the only one of our models I tested this morning. I make no guarantees about it at the moment.

```toml
[dependencies]
aws-sdk-dynamodb = { git = "https://github.com/awslabs/aws-sdk-rust", branch = "main" }
serde = { version = "1", features = ["derive"] }
serde_dynamo = { git = "https://github.com/zenlist/serde_dynamo", branch = "3.0.0-alpha.0" }
tokio = { version = "1", features = ["full"] }
```

```rust
use aws_sdk_dynamodb::Client;
use serde::Deserialize;
use serde_dynamo::aws_sdk_dynamodb::{from_items, to_attribute_value};

type Result<T, E = Box<dyn std::error::Error + Send + Sync>> = std::result::Result<T, E>;

#[tokio::main]
async fn main() -> Result<()> {
    std::env::set_var("AWS_DEFAULT_REGION", "us-east-1");
    std::env::set_var("AWS_ACCESS_KEY_ID", "AK...");
    std::env::set_var("AWS_SECRET_ACCESS_KEY", "...");

    let client = Client::from_env();

    let result = client
        .scan()
        .table_name("things")
        .limit(10)
        .filter_expression("value > :value")
        .projection_expression("id, value")
        .expression_attribute_values(":value", to_attribute_value(3)?)
        .send()
        .await?;

    if let Some(items) = result.items {
        let things = from_items::<Thing>(items)?;
        for thing in things {
            println!("{}: {}", thing.id, thing.value);
        }
    }

    Ok(())
}

#[derive(Debug, Deserialize)]
struct Thing {
    id: String,
    value: u64,
}
```

**Comment by dcormier** [comment]
I did ran into an issue with this. I can't deserialize `AttributeValue::B` to a `serde_json::Value`.

```rust
let av = AttributeValue::B(Blob::new("this text is base64-encoded"));
let js_value: Value = from_attribute_value(av.clone()).unwrap();
```

Fails with:
```text
thread 'aws_sdk_dynamodb::tests::value' panicked at 'called `Result::unwrap()` on an `Err` value: Error(Message("invalid type: byte array, expected any valid JSON value"))', src/aws_sdk_dynamodb/tests.rs:14:14
```

The error is from `generic::de::DeserializerBytes::deserialize_bytes`.

You can see this happen with the tests in #14. It's entirely possible that I'm just doing this wrong.

My original idea was to be able to serialize to [DynamoDB JSON](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.LowLevelAPI.html#Programming.LowLevelAPI.ResponseFormat) and back, and that's what those tests are built around.

**Comment by bryanburgers** [comment]
I just released `3.0.0-alpha.1` of `serde_dynamo` to crates.io, which is more a more complete version of what I was thinking with regards to how to support different dynamodb clients **and** how to support future versions of both `rusoto_dynamodb` and `aws-sdk-dynamodb` without requiring a major version bump every time they bump their pre-1.0 versions.

```toml
[package]
name = "aws_sdk_example"
version = "0.1.0"
edition = "2021"

[dependencies]
aws-config = { version = "0.0.25-alpha" }
aws-sdk-dynamodb = { version = "0.0.25-alpha" }
serde = { version = "1", features = ["derive"] }
serde_dynamo = { version = "3.0.0-alpha.1", features = ["aws-sdk-dynamodb+0_0_25-alpha"] }
tokio = { version = "1", features = ["rt-multi-thread", "macros"] }
```

```rust
use aws_config::meta::region::RegionProviderChain;
use aws_sdk_dynamodb::{Client, Region};
use serde::{Deserialize, Serialize};
use serde_dynamo::from_items;

type Result<T, E = Box<dyn std::error::Error + Send + Sync>> = std::result::Result<T, E>;

#[derive(Debug, Serialize, Deserialize)]
struct User {
    id: UserId,
    first_name: String,
    last_name: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(transparent)]
struct UserId(String);

#[tokio::main]
async fn main() -> Result<()> {
    let region_provider = RegionProviderChain::default_provider().or_else(Region::new("us-west-2"));
    let shared_config = aws_config::from_env().region(region_provider).load().await;
    let client = Client::new(&shared_config);

    let resp = client
        .scan()
        .table_name("users")
        .projection_expression("id, first_name, last_name")
        .send()
        .await?;

    if let Some(items) = resp.items {
        let items: Vec<User> = from_items(items)?;
        println!("{:#?}", items);
    }

    Ok(())
}
```

**Comment by ahornbeck** [comment]
I think there is an issue with the features for `aws_sdk_dynamodb_0_2`.  Or it could be me.

```toml
[dependencies]
aws-config = "0.2.0"
aws-sdk-dynamodb = "0.2.0"
serde_dynamo = { version = "3.0.0-alpha.2", features = ["aws_sdk_dynamodb_0_2"] }
```

```Rust
use aws_sdk_dynamodb::model::AttributeValue;
use serde_dynamo::{to_attribute_value};

let r: AttributeValue = to_attribute_value(String::from("Value")).unwrap();

```
Compiler returns this error.

> the trait `serde_dynamo::AttributeValue` is not implemented for `aws_sdk_dynamodb::model::AttributeValue`

Am I making the mistake or is an implementation missing?  Thx!

**Comment by bryanburgers** [comment]
> I think there is an issue with the features for `aws_sdk_dynamodb_0_2`. Or it could be me.
> 
> ```toml
> [dependencies]
> aws-config = "0.2.0"
> aws-sdk-dynamodb = "0.2.0"
> serde_dynamo = { version = "3.0.0-alpha.2", features = ["aws_sdk_dynamodb_0_2"] }
> ```
> 
> ```rust
> use aws_sdk_dynamodb::model::AttributeValue;
> use serde_dynamo::{to_attribute_value};
> 
> let r: AttributeValue = to_attribute_value(String::from("Value")).unwrap();
> ```
> 
> Compiler returns this error.
> 
> > the trait `serde_dynamo::AttributeValue` is not implemented for `aws_sdk_dynamodb::model::AttributeValue`
> 
> Am I making the mistake or is an implementation missing? Thx!

@ahornbeck The feature you need to enable is called `aws-sdk-dynamodb+0_2`, (not `aws_sdk_dynamodb_0_2`). If you change that, the example program works.

```toml
[dependencies]
aws-config = "0.2.0"
aws-sdk-dynamodb = "0.2.0"
serde_dynamo = { version = "3.0.0-alpha.2", features = ["aws-sdk-dynamodb+0_2"] }
                                                       # ^^^^^^^^^^^^^^^^^^^^
                                                       #        here
```

**Comment by ahornbeck** [comment]
@bryanburgers - so, it was me.  I copied/pasted the module name rather than looking closely at the feature name. Thank you for the help!

**Comment by steveocr** [comment]
Any chance this can be updated to use version 0.3 of the official sdk? Also are there plans to release v3.0.0 soon?

**Comment by noxasaxon** [comment]
Hey @steveocr, as of https://github.com/zenlist/serde_dynamo/pull/19, the alpha branch on crates.io has been updated to the latest aws_sdk version

```toml
[dependencies]
aws-config = {version = "0.4", features=["rustls"]}
aws-types = {version = "0.4"}
aws-sdk-dynamodb = {version = "0.4", features=["rustls"]}
serde_dynamo = { version = "3.0.0-alpha", features = ["aws-sdk-dynamodb+0_4"] }
```

**Comment by bryanburgers** [comment]
Thanks to everyone who has worked to keep the 3.0.0-alpha branch up-to-date!

I just released version 3.0.0 (non-alpha) to crates.io! https://crates.io/crates/serde_dynamo
